posts/loggingPython.md

```diff
@@ -0,0 +1,211 @@
+
+```python
+>>> import sys
+>>>
+>>> def captureException():
+...     return sys.exc_info()
+...
+>>> try:
+...     1/0
+... except:
+...     print(captureException())
+...
+(<class 'ZeroDivisionError'>, ZeroDivisionError('division by zero',), <traceback object at 0x101acb1c8>)
+None
+
+```
+
+![](./images/loggingPython/exception1.png)
+
+While this is useful in smaller scripts, as your application and operation requirements grow, `print` becomes a less viable solution. It doesn't offer you the flexibility to turn off entire categories of output statements and it only allows you to output to the `stdout`. It also misses information such as line number and time that it was generated that can assist with debugging. While `print` is the easiest approach since it doesn't require setup, it can quickly come back to bite you. It's also bad practice to ship a package that prints directly to `stdout` because it removes the user's ability to control the messages.
+
+## logging.info("Hello World to Logging")
+
+The logging module is part of Python's standard library and has been since version 2.3. It automatically adds context, such as line number & timestamps, to your logs. The module makes it easy to add namespace logging and severity levels to give you more control over where and what is output.
+
+I'm a believer that the best way to learn something is to do it, so I'd encourage you to follow along in the Python REPL. Getting started with the logging module is simple, here's all you have to do:
+
+```python
+>>> import logging
+>>> logging.basicConfig()
+>>> logger = logging.getLogger(__name__)
+>>>
+>>> logger.critical('logging is easier than I was expecting')
+CRITICAL:__main__:logging is easier than I was expecting
+```
+
+_What just happened?_ `getLogger` provided us with a logger instance. We then gave it the event 'logging is easier than I was expecting' with a level of `critical`.
+
+### Levels
+
+The Python module allows you to differentiate events based on their severity level. The levels are represented as integers between 0 and 50.  The module defines five constants throughout the spectrum, making it easy to differentiate between messages.
+
+![](./images/loggingPython/loggingLevels.jpeg)
+
+Each level has meaning attached to it, and you should think critically about the level you're logging at.
+
+```python
+>>> logger.critical("this better be bad")
+CRITICAL:root:this better be bad
+>>> logger.error("more serious problem")
+ERROR:root:more serious problem
+>>> logger.warning("an unexpected event")
+WARNING:root:an unexpected event
+>>> logger.info("show user flow through program")
+>>> logger.debug("used to track variables when coding")
+```
+
+By default, the logger will only print `warning`, `error`, or `critical` messages. You can customize this behavior, and even modify it at runtime to activate more verbose logging dynamically.
```

Comment:
I was confused as to why the above 2 lines didn't have output. I would clarify that better here by adding "Notice the `info` and `debug` calls didn't print a message. This is due the logger level defaulting to `warn`, which only prints the `warning`, `error`, and `critical` levels.